nft_ingester/src/backfiller.rs

```diff
@@ -0,0 +1,702 @@
+        Self { prev, curr }
+    }
+}
+
+/// Main struct used for backfiller task.
+struct Backfiller<T: Messenger> {
+    db: DatabaseConnection,
+    listener: PgListener,
+    rpc_client: RpcClient,
+    messenger: T,
+    failure_delay: u64,
+}
+
+impl<T: Messenger> Backfiller<T> {
+    /// Create a new `Backfiller` struct.
+    async fn new(pool: Pool<Postgres>, config: IngesterConfig) -> Self {
+        // Create Sea ORM database connection used later for queries.
+        let db = SqlxPostgresConnector::from_sqlx_postgres_pool(pool.clone());
+
+        // Connect to database using sqlx and create PgListener.
+        let mut listener = sqlx::postgres::PgListener::connect_with(&pool.clone())
+            .await
+            .map_err(|e| IngesterError::StorageListenerError {
+                msg: format!("Could not connect to db for PgListener {e}"),
+            })
+            .unwrap();
+
+        // Get database listener channel.
+        let channel = config
+            .database_config
+            .get(&*DATABASE_LISTENER_CHANNEL_KEY)
+            .and_then(|u| u.clone().into_string())
+            .ok_or(IngesterError::ConfigurationError {
+                msg: format!(
+                    "Database listener channel missing: {}",
+                    DATABASE_LISTENER_CHANNEL_KEY
+                ),
+            })
+            .unwrap();
+
+        // Setup listener on channel.
+        listener
+            .listen(&channel)
+            .await
+            .map_err(|e| IngesterError::StorageListenerError {
+                msg: format!("Error listening to channel on backfill_items tbl {e}"),
+            })
+            .unwrap();
+
+        // Get RPC URL.
+        let rpc_url = config
+            .rpc_config
+            .get(&*RPC_URL_KEY)
+            .and_then(|u| u.clone().into_string())
+            .ok_or(IngesterError::ConfigurationError {
+                msg: format!("RPC URL missing: {}", RPC_URL_KEY),
+            })
+            .unwrap();
+
+        // Instantiate RPC client.
+        let rpc_client = RpcClient::new_with_commitment(rpc_url, CommitmentConfig::processed());
+
+        // Instantiate messenger.
+        let mut messenger = T::new(config.messenger_config).await.unwrap();
+        messenger.add_stream(TRANSACTION_STREAM).await;
+        messenger.set_buffer_size(TRANSACTION_STREAM, 5000).await;
+
+        Self {
+            db,
+            listener,
+            rpc_client,
+            messenger,
+            failure_delay: INITIAL_FAILURE_DELAY,
+        }
+    }
+
+    /// Run the backfiller task.
+    async fn run(&mut self) {
+        loop {
+            match self.get_trees_to_backfill().await {
+                Ok(backfill_trees) => {
+                    if backfill_trees.len() == 0 {
+                        // If there are no trees to backfill, wait for a notification on the channel.
+                        let _notification = self.listener.recv().await.unwrap();
```

Comment:
I could probably make this use a tokio channel that could wrap the database listener channel, that would allow for the more tokio-ish `while let Some(msg) = receiver_channel.recv()` pattern.  But I think the reason I need the top level loop is because really I'm implementing a "do-while" loop: I need to run the backfiller the first time without any notification, and once all data is initially handled, then wait on the channel for any changes to the table.

The fact that you made this comment means maybe the pattern I chose is not clear enough, what do you think?  Should I work on refactoring the loop to be more clear?